spec.emu

```diff
@@ -844,3 +836,109 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
     </emu-clause>
   </emu-clause>
 </emu-clause>
+<emu-clause id="sec-reflection" number=28>
+  <h1>Reflection</h1>
+  <emu-clause id="sec-modulesource-objects">
+    <h1><ins>ModuleSource Objects</ins></h1>
+    <p>Different source record types reflect as different module source class instances in their source phase, which represent a compiled source module which has not yet been linked, instantiated or executed.</p>
+    <p>All ModuleSource reflection Objects are subclasses of %AbstractModuleSource% and provide a [[HostSourceMetadata]] internal slot.</p>
+
+    <emu-clause id="sec-%abstractmodulesource%-intrinsic-object">
+      <h1>The %AbstractModuleSource% Intrinsic Object</h1>
+      <p>The <dfn>%AbstractModuleSource%</dfn> intrinsic object:</p>
+      <ul>
+        <li>is a constructor function object that all of the _AbstractModuleSource_ constructor objects inherit from.</li>
+        <li>along with its corresponding prototype object, provides common properties that are inherited by all _AbstractModuleSource_ constructors and their instances.</li>
+        <li>does not have a global name or appear as a property of the global object.</li>
+        <li>acts as the abstract superclass of the various _AbstractModuleSource_ constructors.</li>
+        <li>will throw an error when invoked, because it is an abstract class constructor. The _AbstractModuleSource_ constructors do not perform a `super` call to it.</li>
+      </ul>
+ 
+      <emu-clause id="sec-%abstractmodulesource%">
+        <h1>%AbstractModuleSource% ( )</h1>
+        <p>This function performs the following steps when called:</p>
+        <emu-alg>
+          1. Throw a *TypeError* exception.
+        </emu-alg>
+        <p>The *"length"* property of this function is *+0*<sub>𝔽</sub>.</p>
+      </emu-clause>
+
+      <emu-note>
+        <p>All subclasses of AbstractModuleSource must provide a @@toStringTag property getter performing an instance branding check (see <emu-xref href="#sec-get-modulesource-prototype-@@tostringtag"></emu-xref>).</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-abstract-operations-for-abstractmodulesource-objects">
+      <h1>Abstract Operations for AbstractModuleSource Objects</h1>
+      <emu-clause id="sec-retrievehostsourcemetadata" type="abstract operation">
+        <h1>
+          RetrieveHostSourceMetadata (
+            _anyModuleSource_: an Object,
+          ): anything
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>Used to retrieve the host-defined data from a subclass of %AbstractModuleSource% to populate for a new instance of its module record by the host.</dd>
+        </dl>
+        <emu-alg>
+          1. Assert: _anyModuleSource_ is an instance of a subclass of %AbstractModuleSource%.
```

Comment:
i'm not sure "instance of a subclass of" is a thing that's defined in the spec.

Reply:
Thanks, removed this terminology.